### ISSUE TITLE
bars.exit().remove() on custom visual doesn't remove bars when interacting with Power BI filters pane

### ISSUE THREAD

**Issue by roseKQ** [Microsoft/PowerBI-visuals-sampleBarChart#8]
Please help -  I'm building a custom visual based on the tutorial here - I've followed it very closely then made tweeks. When I interact with the filters pane though, the visual updates but it doesn't clear - it only clears when I reload the visual. I am using the d3.js methods .exit().remove() but they don't make any difference - any idea what I've missed?? Everything else on the visual seems to be working pretty well. I've appended two screen shots to illustrate what is happening - the first screen shot is it working (I've scribbled out some scores) then the second shot is what happens when I make changes to the filters.

![github1_help](https://cloud.githubusercontent.com/assets/11443407/20751990/bfe9721c-b6f6-11e6-99b9-fa25976df878.PNG)

![github2_help](https://cloud.githubusercontent.com/assets/11443407/20751995/c637e2ac-b6f6-11e6-840f-79f5726a3bec.PNG)

**Comment by amolkare** [comment]
Please clear html container at the beginning of update method 
 document.getElementById("chart").innerHTML = "";
or 
If your are using JQuery $("#chart").html("");

**Comment by roseKQ** [comment]
That just clears the visual completely...

**Comment by roseKQ** [comment]
Here is my update method - I've abbreviated it down - maybe I've missed something here and something else needs to clear? The viewModel? If I've missed something please let me know :,(

`       public update(options: VisualUpdateOptions) {

          
            if(!options.dataViews || !options.dataViews[0])return;

            let width = options.viewport.width;
            let height = options.viewport.height;
            let margin = { top: 250, bottom: 25, left: 150, right: 0 };
            let barHeight = 25;
            let smallBar = 3;

            let viewModel: BarChartViewModel = this.vm = visualTransform(options, this.host);

            let x = d3.scale.linear()
                .domain([0, viewModel.dataMax])
                .range([0, width]);

            this.svg.attr({
                width: width,
                height: height-margin.left
            });

            let bars = this.barContainer.selectAll('.g').data(viewModel.dataPoints);

            bars.enter()
                .append('g')
                .classed('bar', true)
                .attr("transform", function(d,i){ return "translate(450," + i * barHeight + ")"});
                
            bars.append('rect')...
           
            bars.append('text')...

            bars.append("text")...

            bars.append("text") ...

           bars.exit().remove();


    }
`

**Comment by amolkare** [comment]
You have to clear the containers or text containers where you appended the text in update function, do not clear the base containers from Constructor/Init function.
as per above code you probably need to clear bars containers.

**Comment by roseKQ** [comment]
Can you give an example of the code? bars.exit().remove() doesn't do anything,  bars.remove() removes everything completely, is there an inbuilt method I don't know about?? I can't find anything in the documentation :(

**Comment by amolkare** [comment]
Please try this.barContainer.html(""); Similarly all the containers created in Init method i.e. this.axis.html(""); in sample barchart.

**Comment by roseKQ** [comment]
No nothing, the visual just hangs. I'm going to start a new visual from fresh and see if I can work out what the issue is - thanks for your time though

**Comment by roseKQ** [comment]
OK, I've found the problem - in case anyone has a similar issue, I found that the d3.exit().remove() methods DO NOT work on appended group elements; e.g.

`var bars = this.barContainer.selectAll(".g").data(data).enter().append("g")...`

 In order to fix the problem, I had to separate all text, rect, etc. elements into separate containers and then apply the d3.exit().remove() e.g.

`var bars = this.barContainer.selectAll(".bars").data(data).enter().append("rect")...`

It is pretty frustrating that there are issues when using the group element in d3 with power BI but there we are...
